Feature/hype contract

VargaElod23 opened the pull request:
# Generic Pull Request template

## Linked issue

Please link the issue:

ex. Closes #<33>.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [ ] The changes impact front-end.

## For changes impacting frontend

Please leave some screenshots/video recordings of the newly implemented interface and the Figma link for its initial design.

### Screenshots/ recordings

=> Multimedia here.

### Figma link

=> Figma link here.

## Description

Please provide a short description about what the PR contains.

## Funny gif

Leave a funny gif for faster review.

rattrap commented on services/contracts/contracts/DynamicEscrow.sol:
```diff
@@ -0,0 +1,208 @@
+            receiver.transfer(amount);
+        } else {
+            IERC20Upgradeable token = ERC20Upgradeable(_depo.tokenAddress);
+            token.safeTransferFrom(receiver, address(this), amount);
```
I think it's the other way around. From us (`address(this)`) to the receiver.

rattrap commented on services/contracts/contracts/DynamicEscrow.sol:
```diff
@@ -0,0 +1,208 @@
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
```
Anyhow, we don't need `ERC20Upgradeable` and `SafeERC20Upgradeable`, we just need the interface. In this case it's `IERC20Upgradeable`

rattrap commented on services/contracts/contracts/DynamicEscrow.sol:
```diff
@@ -0,0 +1,208 @@
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    using SafeMathUpgradeable for uint256;
+    using AddressUpgradeable for address payable;
+    using CountersUpgradeable for CountersUpgradeable.Counter;
```
We don't use this.

VargaElod23 replied:
Removed

rattrap commented on services/contracts/contracts/interfaces/IEscrow.sol:
```diff
@@ -0,0 +1,33 @@
+        uint256 poolId;
+    }
+
+    struct DepositIndex {
```
I don't think we need a whole structure for this. From the implementation it seems that an poolId => payee => DynamicDeposit is enough

rattrap commented on services/contracts/contracts/interfaces/IRewarder.sol:
```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.8.4;
```
I don't think we use this interface.

VargaElod23 replied:
Yes, this will be used for the rewarder service

rattrap commented on services/contracts/test/hype.test.ts:
```diff
@@ -0,0 +1,48 @@
+    expect(cOwner).to.equal(owner.address);
+  });
+
+  it("Tests pausability", async () => {
```
This just tests the Pausable contract. We need to check if we can create a hype pool if the contract is paused.

VargaElod23 replied:
Added in latest tests

rattrap commented:
So the way the deposit in the escrow contract ties in to the hype pool doesn't seem too safe right now.
You deposit tokens for a non-existing hype pool and then you create the pool. What happens if, after you deposit the tokens, another user comes and creates a pool before you do?

We should first create the pool and then deposit tokens to it making it "active". That's the flow on the web app too.

Elod23 commented on services/contracts/contracts/DynamicEscrow.sol:
```diff
@@ -0,0 +1,208 @@
+        address payee,
+        uint256 poolId,
+        uint256 amount
+    ) public onlyRewarder nonReentrant {
```
Yes, however we manipulate rewards before checking anything, so I would better like to have it here. Deployment gas fees will anyways be irrelevant

VargaElod23 commented:
Hey @rattrap ! So I refactored according to most of the comments you left. The flow is now converted into the three-step one you requested: pool creation -> Escrow lock -> activation. However, this has a security flaw to begin with, but before adding the fix to it, I wanted to discuss it with you.  Right now only the pool checks the escrow, but `To be sure that payments get locked for the right pool and get locked only once` we need to tie the pool contract to the escrow too. For this app this would be fine, but if we'd like to use this contract to accrue rewards for the `community-site` users too, then we'd need such `reward pools` represented as contracts too for the no-fee staking events and so on. Please take a look at the current V and let's ideate on the next iteration of it, what should be added and how the ecosystem should be enlarged.

rattrap commented on services/contracts/contracts/interfaces/IHypePool.sol:
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
```
So we do need it?

VargaElod23 commented on services/contracts/contracts/HypePool.sol:
```diff
@@ -0,0 +1,79 @@
+{
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+
+    CountersUpgradeable.Counter private _tokenIdCounter;
```
Removed to `_poolIds`

rattrap commented on packages/contracts/contracts/HypePool.sol:
```diff
@@ -0,0 +1,125 @@
+            id
+        );
+        require(
+            _deposit.weiAmount == _pool.cap,
```
Should also check that the deposited token address is the same as the one on the pool else someone could deposit DogeCoin and say it's actually USDT

VargaElod23 replied:
Introducing this we couple the two contracts, so if that's needed anyways then  I'll add some checks in other places too for checking of the poolId is correct.

Elod23 commented on packages/contracts/contracts/DynamicEscrow.sol:
```diff
@@ -0,0 +1,214 @@
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
```
Yep, we can do that.

VargaElod23 replied:
Tho I don't think it limits it since in the tests we use a simple ERC20 from OZ: `token/ERC20/ERC20.sol`
